src/main.rs

```diff
@@ -1,3 +1,114 @@
+extern crate gio;
+extern crate gtk;
+
+use gio::prelude::*;
+use gio::MenuExt;
+use gtk::prelude::*;
+
+use std::env::args;
+
+macro_rules! upgrade_weak {
+    ($x:ident, $r:expr) => {{
+        match $x.upgrade() {
+            Some(o) => o,
+            None => return $r,
+        }
+    }};
+    ($x:ident) => {
+        upgrade_weak!($x, ())
+    };
+}
+
+fn build_menu(application: &gtk::Application, window: &gtk::ApplicationWindow) {
+    let menu = gio::Menu::new();
+    let menu_bar = gio::Menu::new();
+    let about_menu = gio::Menu::new();
+
+    menu.append("Quit", "app.quit");
+
+    about_menu.append("About", "app.about");
+    menu_bar.append_submenu("?", &about_menu);
+
+    // Connect the about menu entry click event.
+    let about = gio::SimpleAction::new("about", None);
+    let weak_window = window.downgrade();
+    about.connect_activate(move |_, _| {
+        let window = upgrade_weak!(weak_window);
+        let p = gtk::AboutDialog::new();
+        p.set_authors(&["Sebastian Dröge", "Guillaume Gomez"]);
+        p.set_website_label(Some("github repository"));
+        p.set_website(Some("https://github.com/sdroege/rustfest-rome18-gtk-gst-workshop"));
+        p.set_comments(Some("A process viewer GUI wrote with gtk-rs"));
+        // p.set_copyright(Some("This is under MIT license"));
+        p.set_transient_for(Some(&window));
+        p.set_program_name("Rustfest GTK+GStreamer");
+        p.run();
+        p.destroy();
```

Comment:
I'd recommend not doing a recursive main loop here. Just do `p.show()` and for destroying you can connect to signals on the dialog. See the `listbox` example

Reply:
It's the recommended way to do it.

Reply:
Where does it say so? :)

The problem with this is that you can very easily run into multiple mutable borrows of a `RefCell` for example by doing this. As while the dialog is running, everything (incl all other signal handlers) are called from inside `p.run()`.

Reply:
In the examples but I'll update it if you prefer, I don't really mind. :)

Reply:
See https://github.com/sdroege/rustfest-rome18-gtk-gst-workshop/pull/1#discussion_r221473154

In the context of Rust I'd really recommend to not do recursive main loops. `RefCell` will easily panic on you then if you're not careful, and there's not really any advantage (unlike in C) because we have cheap and ergonomic closures